src/backend/access/external/plugstorage.c

```diff
@@ -0,0 +1,533 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*-------------------------------------------------------------------------
+ *
+ * plugstorage.c
+ *
+ *    Pluggable storage implementation. Support external table for now.
+ *
+ *-------------------------------------------------------------------------
+ */
+#include "access/plugstorage.h"
+#include "catalog/pg_type.h"
+#include "catalog/pg_proc.h"
+#include "catalog/pg_exttable.h"
+#include "nodes/value.h"
+#include "parser/parse_func.h"
+
+/*
+ * getExternalTableTypeList
+ *
+ *    Return the table type for a given external table
+ *
+ *    Currently it is an implementation with performance cost due to the
+ *    fact that the format types for TEXT, CSV, and ORC external tables
+ *    with customer protocol (HDFS) are all 'b' in pg_exttable catalog
+ *    table. It needs to visit the format options to get the table type.
+ */
+void getExternalTableTypeList(const char formatType,
+                              List *formatOptions,
+                              int *formatterType,
+                              char **formatterName)
+{
+	if (fmttype_is_custom(formatType))
+	{
+		Assert(formatOptions);
+
+		*formatterName = getExtTblFormatterTypeInFmtOptsList(formatOptions);
+
+		*formatterType = ExternalTableType_PLUG;
+	}
+	else
+	{
+		*formatterType = ExternalTableType_Invalid;
+		*formatterName = NULL;
+		elog(ERROR, "undefined external table format type: %c", formatType);
+	}
+}
+
+void getExternalTableTypeStr(const char formatType,
```

Comment:
We need some descriptions like the function above, please add more comments before delivery